src/format/kdbx4/dump.rs

```diff
@@ -42,38 +66,47 @@ pub fn dump_kdbx4(
     // derive master key from composite key, transform_seed, transform_rounds and master_seed
     let key_elements: Vec<&[u8]> = key_elements.iter().map(|v| &v[..]).collect();
     let composite_key = crypt::calculate_sha256(&key_elements)?;
-    let transformed_key = header.kdf.get_kdf().transform_key(&composite_key)?;
-    let master_key = crypt::calculate_sha256(&[header.master_seed.as_ref(), &transformed_key])?;
+    let transformed_key = db
+        .settings
+        .kdf_settings
+        .get_kdf(&kdf_seed)
+        .transform_key(&composite_key)?;
+    let master_key = crypt::calculate_sha256(&[&master_seed, &transformed_key])?;
 
     // verify credentials
-    let hmac_key = crypt::calculate_sha512(&[&header.master_seed, &transformed_key, b"\x01"])?;
+    let hmac_key = crypt::calculate_sha512(&[&master_seed, &transformed_key, b"\x01"])?;
     let header_hmac_key = hmac_block_stream::get_hmac_block_key(u64::max_value(), &hmac_key)?;
     let header_hmac = crypt::calculate_hmac(&[&header_data], &header_hmac_key)?;
 
     writer.write(&header_hmac)?;
 
-    let inner_header = match &db.inner_header {
-        InnerHeader::KDBX4(h) => h,
-        _ => return Err(DatabaseSaveError::UnsupportedVersion.into()),
-    };
+    // Initialize inner encryptor from inner header params
+    let mut inner_cipher = db
+        .settings
+        .inner_cipher_suite
+        .get_cipher(&inner_random_stream_key)?;
 
     // dump inner header into buffer
     let mut payload = Vec::new();
-    inner_header.dump(&mut payload)?;
-
-    // Initialize inner decryptor from inner header params
-    let mut inner_cipher = inner_header
-        .inner_random_stream
-        .get_cipher(&inner_header.inner_random_stream_key)?;
+    KDBX4InnerHeader {
+        inner_random_stream: db.settings.inner_cipher_suite.clone(),
+        inner_random_stream_key,
+    }
+    .dump(&db.header_attachments, &mut payload)?;
 
     // after inner header is one XML document
     crate::xml_db::dump::dump(&db, &mut *inner_cipher, &mut payload)?;
 
-    let payload_compressed = header.compression.get_compression().compress(&payload)?;
+    let payload_compressed = db
+        .settings
+        .compression
+        .get_compression()
+        .compress(&payload)?;
 
-    let payload_encrypted = header
-        .outer_cipher
-        .get_cipher(&master_key, header.outer_iv.as_ref())?
+    let payload_encrypted = db
+        .settings
+        .outer_cipher_suite
+        .get_cipher(&master_key, &outer_iv)?
```

Comment:
@sseemayer could we move the IV generation into the `get_cipher` directly? I think this would make it even less error-prone. We could have a `get_cipher` that takes an existing IV (used for parsing), and a `get_fresh_cipher` that only takes a master key and that returns the cipher and a new IV (used for saving).

Reply:
The issue with this is that the `master_key` takes additional steps and inputs to derive, which would make the moved IV generation more complicated.